src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -558,3 +598,238 @@ double TieredHNSWIndex<DataType, DistType>::getDistanceFrom(labelType label,
+template <bool isMultiValue>
+VecSimQueryResult_List
+TieredHNSWIndex<DataType, DistType>::TieredHNSW_BatchIterator::get_current_batch(size_t n_res) {
+    // Set pointers
+    auto bf_res = this->flat_results.results;
+    auto hnsw_res = this->hnsw_results.results;
+    const auto bf_end = bf_res + VecSimQueryResult_Len(this->flat_results);
+    const auto hnsw_end = hnsw_res + VecSimQueryResult_Len(this->hnsw_results);
+
+    // Merge results
+    VecSimQueryResult *batch_res;
+    if (isMultiValue) {
+        batch_res = merge_results<true>(hnsw_res, hnsw_end, bf_res, bf_end, n_res);
+    } else {
+        batch_res = merge_results<false>(hnsw_res, hnsw_end, bf_res, bf_end, n_res);
+    }
+
+    // If we're on a single-value index, update the set of results returned from the FLAT index
+    // before popping them.
+    if (!isMultiValue) {
+        for (auto it = this->flat_results.results; it != bf_res; ++it) {
+            this->returned_results_set.insert(it->id);
+        }
+    }
+
+    // Update results
+    array_pop_front_n(this->flat_results.results, bf_res - this->flat_results.results);
+    array_pop_front_n(this->hnsw_results.results, hnsw_res - this->hnsw_results.results);
+
+    // If we're on a multi-value index, update the set of results returned (from `batch_res`), and
+    // clean up the results
+    if (isMultiValue) {
+        // Update set of results returned
+        for (size_t i = 0; i < array_len(batch_res); ++i) {
+            this->returned_results_set.insert(batch_res[i].id);
+        }
+        // On multi-value indexes, one (or both) results lists may contain results that are already
+        // returned form the other list (with a different score). We need to filter them out.
+        filter_irrelevant_results(this->flat_results);
+        filter_irrelevant_results(this->hnsw_results);
+    }
+
+    // Return current batch
+    return {batch_res, VecSim_QueryResult_OK};
+}
+
+template <typename DataType, typename DistType>
+void TieredHNSWIndex<DataType, DistType>::TieredHNSW_BatchIterator::filter_irrelevant_results(
+    VecSimQueryResult_List &rl) {
+    // Filter out results that were already returned from the FLAT index
+    auto it = rl.results;
+    auto end = it + VecSimQueryResult_Len(rl);
+    // Skip results that not returned yet
+    while (it != end && this->returned_results_set.count(it->id) == 0) {
+        ++it;
+    }
+    // If none of the results were returned from the FLAT index, return
+    if (it == end) {
+        return;
+    }
+    // Mark the current result as the first result to be filtered
+    auto cur_end = it;
+    ++it;
+    // "Append" all results that were not returned from the FLAT index
+    while (it != end) {
+        if (this->returned_results_set.count(it->id) == 0) {
+            *cur_end = *it;
+            ++cur_end;
+        }
+        ++it;
+    }
+    // Update number of results
+    array_hdr(rl.results)->len = cur_end - rl.results;
```

Comment:
no, because the duplications can appear anywhere in the array, not just in the front.
Alternatively, we can add an array API for filtering results from an array if they appear in a given set (essentially moving this logic to the array)

Reply:
I think you should